designs/2018-relative-package-loading/README.md

```diff
@@ -0,0 +1,282 @@
+
+However, with that solution the user has little recourse if two shareable configs both depend on the same plugin, resulting in a "dependency hell" scenario where many pairs of shareable configs would be incompatible with each other due to different dependency versions.
+
+### Use plugins that pull rules from other plugins, instead of shareable configs
+
+[eslint/eslint#3458 (comment)](https://github.com/eslint/eslint/issues/3458#issuecomment-257161846) proposed solving the duplicate-name problem by using plugins that depend on other plugins and reexport the rules of their dependencies, without any changes to ESLint core. It suggested two possible ways of re-exporting the rules: either a plugin could export them directly with the same name, or it could give the names a common prefix. This was proposed to address the issue that end users are exposed to their configs' dependencies.
+
+Unfortunately, this solution has a few downsides that prevented it from being widely adopted:
+
+* It would require everyone using shareable configs to switch to using plugins, which was regarded by some config authors as an unacceptably large breaking change.
+* It could create confusion about where a rule came from (and where bugs should be reported), because a rule might have been passed through many different plugins before reaching the user.
+* It still caused issues if two loaded plugins exported rules with the same name, resulting in either (a) a naming conflict where one rule would be unconfigurable, or (b) scoped rule names where the end user would end up exposed to their config's dependencies anyway.
+* This solution would not fix the issue that ESLint depends on package managers' implementation details (at least partly because that issue was not known at the time that the solution was proposed).
+
+### Recommend that shareable configs install their peer dependencies with a postinstall script
+
+Another possible course of action would be to encourage shareable configs to install their peer dependencies with a `postinstall` script. This would avoid adding complexity to ESLint core, while preventing users from needing to manually install peer dependencies.
+
+Unfortunately, this would cause a few issues:
+
+* If two shareable configs depended on different versions of the same plugin, their postinstall scripts would conflict with each other, resulting in one of the shareable configs having an incompatible peerDependency. This would require end users to manually remove a shareable config, and end users might end up confused about why they can't use two shareable configs together.
+* With this solution, ESLint would continue to depend on implementation details of package managers, so it would still break under some valid setups (e.g. in `lerna` monorepos).
```

Comment:
I think ESLint is already tightly coupled to how npm works, so I don't think this is a big concern.

Reply:
I strongly disagree. I think there's an important distinction between:

* Coupling with the de-facto `package.json` spec (i.e. assuming that if someone specifies a package in `dependencies` and calls `require` on it, the package will be found). This is a reasonable place to be.
* Coupling with the specific implementation details of how a package manager implements the `package.json` spec (i.e. assuming that if we call `require` on something, it will be found because *someone else* has a dependency on it). This is where we are now, and it has caused major problems where ESLint is sometimes inherently unable to load the user's packages if they're installed with Yarn Plug n' Play or Lerna. (For the full details on why this doesn't work, see https://github.com/eslint/eslint/issues/10125.)

    This isn't just a distinction between npm and other package managers -- ESLint can also break or behave unexpectedly under certain circumstances even with npm, because we're not relying on supported behavior. (To illustrate: suppose the user decides they want to use `espree@3` for some reason, so they add `espree@3` as a dependency and include `parser: espree` in their config. Their code will still be parsed with `espree@4` because ESLint happens to depend on that package, and ESLint loads the user's parser from its own dependencies rather than the dependencies that the user can control.)

Reply:
I think it's appropriate to couple to how *node* works, however - and if the way some tools install packages breaks that, that's *their* breakage, and isn't something eslint necessarily needs to cleave to.

Reply:
I agree in the abstract, but I don't think that case is applicable here. Node's behavior depends on how packages are arranged in the filesystem, and the package manager is in charge of arranging the packages in the filesystem. `package.json` files are the way we express constraints on a package manager.

So:

* A package manager is correct if it arranges packages in a manner such that Node will able to load all the dependency relationships expressed in the `package.json` files.
* A library is correct (as far as dependencies are concerned) if it accurately expresses its dependency relationships in the appropriate `package.json` file(s).

The result is that if both the library and the package manager are correct, then Node will be able to load the library's dependencies. In this case, ESLint has dependencies that aren't expressed in a `package.json` file (namely, it needs to load user-provided plugins as if they were ESLint's own dependencies). So the bug is in ESLint, not the package manager.